src/test/test.rs

```diff
@@ -336,60 +342,66 @@ enum PathCondition {
 
 #[cfg(not(windows))]
 fn path(path: &[u8], cond: PathCondition) -> bool {
-    use libc::{lstat, stat, S_IFBLK, S_IFCHR, S_IFDIR, S_IFLNK, S_IFMT, S_IFREG};
-    use libc::{mode_t, S_IFIFO};
-    use std::ffi::CString;
+    use std::os::unix::fs::{MetadataExt, FileTypeExt};
+    use std::os::unix::ffi::OsStrExt;
+    use std::fs::{self, Metadata};
+    use std::ffi::OsStr;
 
-    static S_ISUID: mode_t = 0o4000;
-    static S_ISGID: mode_t = 0o2000;
-    static S_IFSOCK: mode_t = 0o140000;
+    let path = OsStr::from_bytes(path);
+
+    static S_ISUID: u32 = 0o4000;
+    static S_ISGID: u32 = 0o2000;
 
     enum Permission {
         Read = 0o4,
         Write = 0o2,
         Execute = 0o1,
     }
-    let perm = |stat: stat, p: Permission| {
-        use libc::{getgid, getuid};
-        let (uid, gid) = unsafe { (getuid(), getgid()) };
-        if uid == stat.st_uid {
-            stat.st_mode & ((p as mode_t) << 6) != 0
-        } else if gid == stat.st_gid {
-            stat.st_mode & ((p as mode_t) << 3) != 0
+
+    let perm = |metadata: Metadata, p: Permission| {
+        #[cfg(not(target_os = "redox"))]
+        let (uid, gid) = unsafe { (libc::getuid(), libc::getgid()) };
+        #[cfg(target_os = "redox")]
+        let (uid, gid) = (syscall::getuid().unwrap() as u32,
```

Comment:
Is it possible for either of these to fail?  I can’t really think of a scenario where it would be but I just want to make sure.

Reply:
It shouldn't be possible; looking at the [implementation](https://github.com/redox-os/kernel/blob/master/src/syscall/privilege.rs), it should only fail if the calling process doesn't exist, which obviously shouldn't happen (or execution shouldn't continue).

@jackpot51 Perhaps these functions shouldn't return `Result`s?

Reply:
Okay, thanks for checking.